L79-cpp-async-response-reader-destruction.md

```diff
@@ -0,0 +1,123 @@
+correct: `std::unique_ptr<ClientAsyncResponseReader>` doesn't actually need to
+do anything. But:
+
+*   Of course the objects may be deleted other ways. If I have a
+    `std::unique_ptr<ClientAsyncResponseReaderInterface>` (i.e. without a
+    specific deleter type), it would be natural for me to assume that I can do
+    `delete p.release()` to get the same effect as destroying the unique
+    pointer. But that is not true, and doing so will be undefined behavior.
+
+    So from this point of view `dd36b153` is not correct, or at least makes for
+    a surprising API.
+
+*   `ClientAsyncResponseReader` is not the only subclass of the interface. The
+    interface's destructor is virtual, so other subclasses may have destructors
+    with side effects. However they are never actually run! This is a huge
+    surprise: it took me hours yesterday to figure out that I wasn't actually
+    crazy: `delete p.release()` really was working differently than just letting
+    a `std::unique_ptr p` go out of scope.
+
+    So from this point of view `e8a61d63` is also not correct.
+
+This is not a theoretical concern: I've found multiple hacks in user
+code working around this with a grumbling tone. For example, see
+[this hack](https://github.com/googleapis/google-cloud-cpp/blob/74de20227371119ea8b8c788d37ae4e24be8ed5d/google/cloud/bigtable/testing/mock_response_reader.h#L77-L106)
+in the cloud bigtable library that uses the phrase "terrible, horrible, no
+good, very bad hack" that involves managing a
+`ClientAsyncResponseReaderInterface` object with **two** `unique_ptr`s (making
+them not so unique).
+
+## Proposal
+
+My plan is split into two commits:
+
+1.  Add a `virtual void Destroy()` method to
+    `ClientAsyncResponseReaderInterface`, and update the `std::default_delete`
+    specializations to call it. In this commit there are two implementations of
+    the `Destroy` method:
+
+    *   The interface itself implements the method to do nothing, with a link
+        to the explanation of the history for why it does nothing.
+
+    *   `ClientAsyncResponseReader` also implements the method to do nothing,
+        with an explanation that it knows it is allocated on an arena and has
+        historically never run its destructor.
+
+    Thus with this commit there is no behavior change for anybody, except in
+    one edge case: if a hack similar to the one mentioned above didn't ensure
+    that the `unique_ptr` to the interface didn't outlive the unique pointer to
+    the concrete subclass, then the former's call to `Destroy` would be
+    undefined behavior after the latter destroyed the object itself (and its
+    vtable). This is unfortunate, but not likely to be common, and the whole
+    point of this proposal is to make such hacks unnecessary.
+
+2.  Change the interface's implementation of the `Destroy` method to do `delete
+    this`, a more suitable default behavior that frees future subclasses from
```

Comment:
This can break some random subclass that was relying on the dtor being a no-op.  I believe it needs a major gRPC version bump when we merge the commit.

We should clarify that here in the proposal.

Reply:
That is in theory also true of the change above; see the last paragraph. Does it need a major bump? Are major version bumps actually done, or does that make this a no-go?

Reply:
We call "major version" bump as really going from gRPC 1.x to 1.x+1 not gRPC 2. Yes this is something that's done often and there is no issue doing so, once the gRFC is approved.

Reply:
Okay, I added text saying each should result in a version bump. PTAL.